### ISSUE TITLE
Need to generate proper error message upon API call between client and server with incompatible API version

### ISSUE THREAD

**Issue by ashutshkumr** [open-traffic-generator/openapiart#389]
**Version information**
v0.2.6

**Describe the bug**
Whenever an API call is made from a client to server with incompatible API versions, a proper error message is not reported back to client highlighting the root cause.

**To Reproduce**
Pick a gosnappi client based on models version X and ixia-c based on models version Y, where X supports certain features not supported in Y.

Next step is to execute gosnappi test exercising aforementioned feature against ixia-c.

**Expected behavior**
A clear and concise error message needs to be reported back as a response whenever API version incompatibility is detected.
- It shouldn't break existing tests
- There shouldn't be a need to add any extra code in tests by default
- This needs to be a default behavior but users should be able to opt out of it (needed during development phase)

**Comment by ashutshkumr** [comment]
# Proposal

**TL;DR:** _openapiart_ would automatically generate business logic for API-compatibility-check between client and server, _auto-initiated_ upon first API call

### Auto-generate API endpoints

1. Following HTTP API shall be auto-generated for all YAML spec bundled using _openapiart_.
   ```bash
   HTTP Method: GET
   HTTP URI: /capabilities/version
   HTTP Response: {"api_version": "2.1.0", "server_version": "2.3.1" }
   ```

2. Corresponding RPC and message shall, by design, be automatically generated for .proto generated from aforementioned YAML bundled using _openapiart_ .

3. SDK providers shall be able to opt out of auto-generating aforementioned API endpoint like so:
   ```python
   import openapiart

   openapiart.OpenApiArt(
       api_files=[".api.yaml"],
       generate_version_api=False
   )
   ```

### Auto-generate API handler in server SDK

1. The client/server SDK for both Go and Python shall have following method on API handle auto-generated
   ```go
   api := gosnappi.NewApi()

   // return a version string specified in original API YAML spec
   api.ApiVersion()

   // shall be set during server init
   api.SetServerVersion()
   // so that following returns a valid server version string (otherwise will return an empty string)
   api.ServerVersion()
   ```

2. The server SDK for both Go and Python shall have a gRPC and HTTP handler for `GET /capabilities/version` which returns
   ```bash
   # if server explicitly sets the server version during init
   {"api_version": "2.1.0", "server_version": "2.3.1" }
   # if server does not explicitly set the server version
   {"api_version": "2.1.0", "server_version": "" }
   ```

### Auto-generate API call for version check in client SDK

1. The client SDK for both Go and Python shall, by design, have following method on API handle auto-generated
   ```go
   api := gosnappi.NewApi()
   api.NewHttpTransport().SetLocation("https://localhost:8443")

   // makes a remote call to server and fetches API and server version, regardless of transport
   api.GetVersion()
   ```

2. The client SDK for both Go and Python shall additionally have following method on API handle auto-generated
   ```go
   api := gosnappi.NewApi()
   api.NewHttpTransport().SetLocation("https://localhost:8443")

   v, _ := api.GetVersion()
   // check compatibility between client API version and server API version, using semantic versioning scheme
   // err should only hold errors that are not related to version check
   compatible, err := api.CheckVersionCompatibility(api.Version(), v.ApiVersion())
   ```

3. End users shall ideally never have to manually invoke `api.CheckVersionCompatibility`. Instead, it shall be automatically invoked upon first ever API call, which is recursively not an `api.GetVersion` call.
   a. If `err` is not `nil` on first version check, the API call shall fail and the check shall be performed again in consecutive API calls
   b. If `err` is `nil` but `compatible` is `false` on first version check, the API call shall fail with user-friendly error and shall be cached (so that consecutive API calls return the same error)
   b. If `err` is `nil` and `compatible` is `true` on first version check, there won't be further checks in consecutive API calls
   ```go
   apiA := gosnappi.NewApi()
   apiA.NewHttpTransport().SetLocation("https://localhost:8443")

   // this will do version check
   apiA.SetConfig()
   // this won't do version check if previous step resulted in successful version check (regardless of whether SetConfig API failed)
   apiA.GetMetrics()

   apiB := gosnappi.NewApi()
   apiB.NewHttpTransport().SetLocation("https://localhost:8443")

   // this will do version check again, since the API handle is new
   apiB.SetConfig()
   // this won't do version check if previous step resulted in successful version check (regardless of whether SetConfig API failed)
   apiB.GetMetrics()
   ```

4. End-users shall be able to opt out of automatic API version compatibility check like so:
   ```go
   api := gosnappi.NewApi()
   api.SetApiVersionCompatibilityCheck(false)
   ```

### Backwards Compatibility

This change shall not require changes in any existing tests.

### Scope

This change shall currently only affect following communication channels:
1. `snappi -> ixia-c-controller`
2. `gnmi-server -> ixia-c-controller`

### Limitations

1. An extra call for version check might affect API call duration, for first ever such call.

2. API version compatibility check won't be supported if an end-user is making API calls using a tooling not generated using _openapiart_.  Although, end-users shall still be able to manually fetch version from server.

3. To support communication channel `ixia-c-controller -> ixia-c-protocol-engine`, manual implementation on server is needed

4. To support communication channel `ixia-c-traffic-engine -> ixia-c-protocol-engine`, manual implementation on both client and server is needed

5. To support communication channel `ixia-c-controller -> ixia-c-traffic-engine`, manual API exposure and implementation on both client and server is needed
